pkg/controller/elasticsearch/driver/upscale_state.go

```diff
@@ -101,41 +122,102 @@ func (s *upscaleState) recordMasterNodeCreation() {
 	if s.isBootstrapped {
 		s.allowMasterCreation = false
 	}
+	s.recordNodesCreation(1)
 }
 
-func (s *upscaleState) canAddMasterNode() bool {
-	return s.allowMasterCreation
+func (s *upscaleState) canCreateMasterNode() bool {
+	return s.getMaxNodesToCreate(1) == 1 && s.allowMasterCreation
 }
 
-func (s *upscaleState) limitMasterNodesCreation(
-	actualStatefulSets sset.StatefulSetList,
-	ssetToApply appsv1.StatefulSet,
-) appsv1.StatefulSet {
-	if !label.IsMasterNodeSet(ssetToApply) {
-		return ssetToApply
+func (s *upscaleState) recordNodesCreation(count int32) {
+	s.recordedCreates += count
+}
+
+func (s *upscaleState) getMaxNodesToCreate(noMoreThan int32) int32 {
+	if s.createsAllowed == nil {
+		// unbounded, so allow all that was requested
+		return noMoreThan
 	}
 
-	targetReplicas := sset.GetReplicas(ssetToApply)
-	actual, alreadyExists := actualStatefulSets.GetByName(ssetToApply.Name)
-	actualReplicas := int32(0)
-	if alreadyExists {
-		actualReplicas = sset.GetReplicas(actual)
+	left := *s.createsAllowed - s.recordedCreates
+	if left < noMoreThan {
+		return left
+	}
+
+	return noMoreThan
+}
+
+func (s *upscaleState) limitNodesCreation(
+	actual appsv1.StatefulSet,
+	toApply appsv1.StatefulSet,
+) (appsv1.StatefulSet, error) {
+	if label.IsMasterNodeSet(toApply) {
+		return s.limitMasterNodesCreation(actual, toApply)
+	}
+
+	actualReplicas := sset.GetReplicas(actual)
+	targetReplicas := sset.GetReplicas(toApply)
+	if actualReplicas == targetReplicas {
+		return toApply, nil
+	}
+
+	if err := s.buildOnce(); err != nil {
+		return appsv1.StatefulSet{}, err
+	}
+	nodespec.UpdateReplicas(&toApply, common.Int32(actualReplicas))
+	replicasToCreate := targetReplicas - actualReplicas
+	replicasToCreate = s.getMaxNodesToCreate(replicasToCreate)
+	if replicasToCreate > 0 {
+		nodespec.UpdateReplicas(&toApply, common.Int32(actualReplicas+replicasToCreate))
+		s.recordNodesCreation(replicasToCreate)
+		ssetLogger(toApply).Info(
+			"Creating nodes",
+			"actualReplicas", actualReplicas,
+			"replicasToCreate", replicasToCreate,
+		)
+	} else {
+		ssetLogger(toApply).Info(
+			"Limiting nodes creation to respect MaxSurge setting",
+			"target", targetReplicas,
+			"actual", actualReplicas,
+		)
+	}
+
+	return toApply, nil
+}
+
+func (s *upscaleState) limitMasterNodesCreation(
+	actual appsv1.StatefulSet,
+	toApply appsv1.StatefulSet,
+) (appsv1.StatefulSet, error) {
+	actualReplicas := sset.GetReplicas(actual)
+	targetReplicas := sset.GetReplicas(toApply)
+
+	if actualReplicas < targetReplicas {
+		if err := s.buildOnce(); err != nil {
```

Comment:
too bad we have to init the state at 2 different places. Can we do it at a single place?

Reply:
Yes :)